File: server/Simulation/Components/EngineComponent.cs
using System.Runtime.InteropServices;
using server.ECS;
using server.Enums;

namespace server.Simulation.Components;

/// <summary>
/// Engine component controlling ship propulsion, throttle, and RCS state.
/// Manages thrust force generation and energy consumption.
/// </summary>
[Component(ComponentType = ComponentType.Engine, NetworkSync = true)]
[StructLayout(LayoutKind.Sequential, Pack = 1)]
public struct EngineComponent(float maxThrustNewtons)
{
    /// <summary>
    /// Tick when this component was last changed, used for network sync.
    /// MUST be first field for raw byte access in ComponentSerializer.
    /// </summary>
    public long ChangedTick = NttWorld.Tick;

    /// <summary>Energy consumption rate based on thrust capacity</summary>
    public float PowerUse = maxThrustNewtons * 0.01f;
    /// <summary>Current throttle setting (0 to 1)</summary>
    public float Throttle = 0;
    /// <summary>Maximum thrust force in Newtons</summary>
    public float MaxThrustNewtons = maxThrustNewtons;
    /// <summary>Whether RCS (Reaction Control System) is active for dampening</summary>
    public bool RCS = true;
}

File: pixiejsClient/src/app/ecs/components/EngineComponent.ts
import { Component, component, serverField } from "../core/Component";
import { ServerComponentType as ServerComponentType } from "../../enums/ComponentIds";

export interface EngineConfig {
  maxPropulsion: number;
  powerUse?: number;
  throttle?: number;
  rcs?: boolean;
}

@component(ServerComponentType.Engine)
export class EngineComponent extends Component {
  // changedTick is inherited from Component base class
  @serverField(1, "f32") powerUse: number; // MaxPropulsion * 0.01
  @serverField(2, "f32") throttle: number; // 0.0 to 1.0
  @serverField(3, "f32") maxPropulsion: number; // Maximum thrust force in Newtons
  @serverField(4, "bool") rcs: boolean; // Reaction Control System

  constructor(entityId: string, config?: EngineConfig) {
    super(entityId);

    if (config) {
      this.maxPropulsion = config.maxPropulsion;
      this.powerUse = config.powerUse !== undefined ? config.powerUse : this.maxPropulsion * 0.01;
      this.throttle = config.throttle || 0;
      this.rcs = config.rcs !== undefined ? config.rcs : true;
    } else {
      // Defaults for deserialization
      this.maxPropulsion = 0;
      this.powerUse = 0;
      this.throttle = 0;
      this.rcs = true;
    }
  }
}
